### ISSUE TITLE
Highstate formatter ignores --state-output

### ISSUE THREAD

**Issue by daemonkeeper** [saltstack/salt#37114]
After upgrading from 2016.3.2 to 2016.3.3 I noticed that the state-output filter for salt appears to be broken. All commands below run on the same machine, targeting the same minion and using the same master (and minion) configuration. 

In Salt 2016.3.2 things seem to work as I'd expect:

    [me@salt-master:~] $ salt --versions-report
    Salt Version:
            Salt: 2016.3.2
    
    Dependency Versions:
            cffi: 0.8.6
        cherrypy: Not Installed
        dateutil: 2.2
            gitdb: 0.6.4
        gitpython: 1.0.1
            ioflo: Not Installed
            Jinja2: 2.7.3
            libgit2: Not Installed
            libnacl: Not Installed
        M2Crypto: 0.21.1
            Mako: 1.0.0
    msgpack-pure: Not Installed
    msgpack-python: 0.4.2
    mysql-python: 1.2.3
        pycparser: 2.10
        pycrypto: 2.6.1
            pygit2: Not Installed
            Python: 2.7.9 (default, Jun 29 2016, 13:08:31)
    python-gnupg: Not Installed
            PyYAML: 3.11
            PyZMQ: 14.4.0
            RAET: Not Installed
            smmap: 0.9.0
            timelib: Not Installed
            Tornado: 4.2.1
                ZMQ: 4.0.5
    
    System Versions:
            dist: debian 8.6 
            machine: x86_64
            release: 3.16.0-4-amd64
            system: Linux
            version: debian 8.6 
    
    [me@salt-master:~] $ salt --state-output=full  $MINIONTARGET$ state.highstate                                                                                           
    jid: 20161020124454202686
    $MINIONTARGET$:
    ----------
            ID: net.ipv4.tcp_keepalive_intvl
        Function: sysctl.present
        Result: True
        Comment: Sysctl value net.ipv4.tcp_keepalive_intvl = 500 is already set
        Started: 12:22:24.732317
        Duration: 25.047 ms
        Changes:   
        ...

and with `--state-output=changes`:

    [me@salt-master:~] $ salt --state-output=changes  $MINIONTARGET$ state.highstate
    jid: 20161020124603538212
    $MINIONTARGET$:
    Name: net.ipv4.tcp_keepalive_intvl - Function: sysctl.present - Result: Clean
    ....

Now, after upgrading to 2016.3.3 this happens:

    [me@salt-master:~] 1 $ salt --versions-report
    Salt Version:
            Salt: 2016.3.3
    
    Dependency Versions:
            cffi: 0.8.6
        cherrypy: Not Installed
        dateutil: 2.2
            gitdb: 0.6.4
        gitpython: 1.0.1
            ioflo: Not Installed
            Jinja2: 2.7.3
            libgit2: Not Installed
            libnacl: Not Installed
        M2Crypto: 0.21.1
            Mako: 1.0.0
    msgpack-pure: Not Installed
    msgpack-python: 0.4.2
    mysql-python: 1.2.3
        pycparser: 2.10
        pycrypto: 2.6.1
            pygit2: Not Installed
            Python: 2.7.9 (default, Jun 29 2016, 13:08:31)
    python-gnupg: Not Installed
            PyYAML: 3.11
            PyZMQ: 14.4.0
            RAET: Not Installed
            smmap: 0.9.0
            timelib: Not Installed
            Tornado: 4.2.1
                ZMQ: 4.0.5
    
    System Versions:
            dist: debian 8.6 
            machine: x86_64
            release: 3.16.0-4-amd64
            system: Linux
            version: debian 8.6 

    ⌂107% [me@salt-master:~] $ salt --state-output=full  $MINIONTARGET$ state.highstate   
    jid: 20161020124739315661
    $MINIONTARGET$:
    ----------
            ID: net.ipv4.tcp_keepalive_intvl
        Function: sysctl.present
        Result: True
        Comment: Sysctl value net.ipv4.tcp_keepalive_intvl = 500 is already set
        Started: 12:25:09.645919
        Duration: 25.79 ms
        Changes:   
    ----------

So far all good, `--state-output=full` works as expected. However, using `--state-output=changes` I suddently get this output:

    [me@salt-master:~] $ salt --state-output=changes  $MINIONTARGET$ state.highstate
    jid: 20161020124848799601

        sysctl_|-net.ipv4.tcp_keepalive_intvl_|-net.ipv4.tcp_keepalive_intvl_|-present:
            ----------
            __id__:
                net.ipv4.tcp_keepalive_intvl
            __run_num__:
                0
            changes:
                ----------
            comment:
                Sysctl value net.ipv4.tcp_keepalive_intvl = 500 is already set
            duration:
                25.696
            name:
                net.ipv4.tcp_keepalive_intvl
            result:
                True
            start_time:

This looks like the --out=yaml outputter. I have no clue why it would use this syntax. Using `--state-output=terse` makes no difference either.  Using `--out=yaml`  indeed gives the same results. I wonder where the yaml formatter would come from. 

Hence I tried to force the formatter with `salt --out=highstate --state-output=changes` which gives exactly the same results. Indeed I tried to quickly find out if the state-output variable is correctly passed. Thus, I added 

    print("OUTPUT: %s" % __opts__.get('state_output'))

to salt/output/highstate.py's `_format_host()` function. However, this gives the expected results:

    [me@salt-master:~] $ salt --state-output=changes  $MINIONTARGET$ state.highstate
    jid: 20161020125423806326
    OUTPUT: changes
    $MINIONTARGET$:
    ...

    [me@salt-master:~] $ salt --out=yaml  $MINIONTARGET$ state.highstate            
    jid: 20161020125522790126
    $MINIONTARGET$:
    ...

So, Salt indeed routes the request to the highstate formatter, and indeed passes the correct arguments. I lack the knowledge of Salt internals to make further sense of the internals. But apparently the highstate format-outputter somehow broke between the releases. The diff doesn't give me any clue, so the issue is probably somewhere else:

```diff
git diff v2016.3.2 v2016.3.3 -- salt/output/highstate.py 
diff --git a/salt/output/highstate.py b/salt/output/highstate.py
index 65916fa..cd7e50e 100644
--- a/salt/output/highstate.py
+++ b/salt/output/highstate.py
@@ -140,6 +140,8 @@ def output(data):
         'Data passed to highstate outputter is not a valid highstate return: %s',
         data
     )
+    # We should not reach here, but if we do return empty string
+    return ''
 
 
 def _format_host(host, data):
@@ -208,7 +210,8 @@ def _format_host(host, data):
                               .format(ret.get('duration', 0)))
 
             tcolor = colors['GREEN']
-            schanged, ctext = _format_changes(ret['changes'])
+            orchestration = ret.get('__orchestration__', False)
+            schanged, ctext = _format_changes(ret['changes'], orchestration)
             nchanges += 1 if schanged else 0
 
             # Skip this state if it was successful & diff output was requested
@@ -443,7 +446,7 @@ def _format_host(host, data):
                                                line_max_len - 7)
         hstrs.append(colorfmt.format(colors['CYAN'], totals, colors))
 
-        if __opts__.get('state_output_profile', False):
+        if __opts__.get('state_output_profile', True):
             sum_duration = sum(rdurations)
             duration_unit = 'ms'
             # convert to seconds if duration is 1000ms or more
@@ -461,15 +464,37 @@ def _format_host(host, data):
     return u'\n'.join(hstrs), nchanges > 0
 
 
-def _format_changes(changes):
+def _nested_changes(changes):
     '''
-    Format the changes dict based on what the data is
+    Print the changes data using the nested outputter
     '''
     global __opts__  # pylint: disable=W0601
 
+    opts = __opts__.copy()
+    # Pass the __opts__ dict. The loader will splat this modules __opts__ dict
+    # anyway so have to restore it after the other outputter is done
+    if __opts__['color']:
+        __opts__['color'] = u'CYAN'
+    __opts__['nested_indent'] = 14
+    ret = u'\n'
+    ret += salt.output.out_format(
+            changes,
+            'nested',
+            __opts__)
+    __opts__ = opts
+    return ret
+
+
+def _format_changes(changes, orchestration=False):
+    '''
+    Format the changes dict based on what the data is
+    '''
     if not changes:
         return False, u''
 
+    if orchestration:
+        return True, _nested_changes(changes)
+
     if not isinstance(changes, dict):
         return True, u'Invalid Changes data: {0}'.format(changes)
 
@@ -483,18 +508,7 @@ def _format_changes(changes):
             changed = changed or c
     else:
         changed = True
-        opts = __opts__.copy()
-        # Pass the __opts__ dict. The loader will splat this modules __opts__ dict
-        # anyway so have to restore it after the other outputter is done
-        if __opts__['color']:
-            __opts__['color'] = u'CYAN'
-        __opts__['nested_indent'] = 14
-        ctext = u'\n'
-        ctext += salt.output.out_format(
-                changes,
-                'nested',
-                __opts__)
-        __opts__ = opts
+        ctext = _nested_changes(changes)
     return changed, ctext
 
 
@@ -516,7 +530,7 @@ def _format_terse(tcolor, comps, ret, colors, tabular):
                 c=colors, w='\n'.join(ret['warnings'])
             )
         fmt_string += u'{0}'
-        if __opts__.get('state_output_profile', False):
+        if __opts__.get('state_output_profile', True):
             fmt_string += u'{6[start_time]!s} [{6[duration]!s} ms] '
         fmt_string += u'{2:>10}.{3:<10} {4:7}   Name: {1}{5}'
     elif isinstance(tabular, str):
@@ -528,7 +542,7 @@ def _format_terse(tcolor, comps, ret, colors, tabular):
                 c=colors, w='\n'.join(ret['warnings'])
             )
         fmt_string += u' {0} Name: {1} - Function: {2}.{3} - Result: {4}'
-        if __opts__.get('state_output_profile', False):
+        if __opts__.get('state_output_profile', True):
             fmt_string += u' Started: - {6[start_time]!s} Duration: {6[duration]!s} ms'
         fmt_string += u'{5}'
```

**Comment by Ch3LL** [comment]
@daemonkeeper I believe you are running into the same issue as #35989 which has been fixed in #36028

can you give that fix a try and report back whether that works?

**Comment by daemonkeeper** [comment]
Yes it does! Thanks for the pointer. I've cherry-picked b9fc51a1727a3b434291acdd2b7d78692731431f on top of the release branch, and it fixes the issue I'm facing.
